symbolic_regression/src/bench.rs

```diff
@@ -147,3 +151,428 @@ pub fn run_constant_opt_linear(env: &ConstantOptLinearEnv) -> (bool, f64, Vec<f6
+        x.push(v);
+    }
+    let mut y = Vec::with_capacity(n_rows);
+    for _ in 0..n_rows {
+        let v: f32 = rng.sample(StandardNormal);
+        y.push(v);
+    }
+    Dataset::new(
+        Array2::from_shape_vec((n_features, n_rows), x).unwrap(),
+        Array1::from_vec(y),
+    )
+}
+
+fn make_population(
+    seed: u64,
+    dataset: &Dataset<BenchT>,
+    options: &Options<BenchT, BENCH_D>,
+    pop_size: usize,
+    tree_size: usize,
+) -> (
+    crate::population::Population<BenchT, BenchOps, BENCH_D>,
+    crate::adaptive_parsimony::RunningSearchStatistics,
+) {
+    let mut rng = StdRng::seed_from_u64(seed);
+    let tagged = TaggedDataset::new(dataset, None);
+    let mut evaluator = Evaluator::new(dataset.n_rows);
+
+    let mut members = Vec::with_capacity(pop_size);
+    for _ in 0..pop_size {
+        let expr = random_expr::<BenchOps, BENCH_D, _>(&mut rng, &options.operators, dataset.n_features, tree_size);
+        let mut member = PopMember::from_expr(expr, dataset.n_features, options);
+        let _ = member.evaluate(&tagged, options, &mut evaluator);
+        members.push(member);
+    }
+
+    let mut stats = crate::adaptive_parsimony::RunningSearchStatistics::new(options.maxsize, 100_000);
+    stats.update_from_population(members.iter().map(|m| m.complexity));
+    stats.normalize();
+
+    (crate::population::Population::new(members), stats)
+}
+
+pub struct SearchBenchEnv {
+    datasets: Vec<Dataset<BenchT>>,
+    options: Vec<Options<BenchT, BENCH_D>>,
+}
+
+pub fn search_env() -> SearchBenchEnv {
+    let seeds = [1u64, 2, 3];
+    let datasets: Vec<_> = seeds.iter().map(|&seed| make_dataset(seed, 1_000, 5)).collect();
+    let options: Vec<_> = seeds.iter().map(|&seed| make_search_options(seed)).collect();
+    SearchBenchEnv { datasets, options }
+}
+
+pub fn run_equation_search(env: &SearchBenchEnv) {
+    for (dataset, options) in env.datasets.iter().zip_eq(&env.options) {
+        let _ = crate::equation_search::<BenchT, BenchOps, BENCH_D>(dataset, options);
+    }
+}
+
+pub struct BestOfSampleBenchEnv {
+    rng: Rng,
+    pop: crate::population::Population<BenchT, BenchOps, BENCH_D>,
+    stats: crate::adaptive_parsimony::RunningSearchStatistics,
+    options: Options<BenchT, BENCH_D>,
+}
+
+pub fn best_of_sample_env() -> BestOfSampleBenchEnv {
+    let options = make_utils_options();
+    let dataset = make_random_dataset(0, 32, 1);
+    let (population, stats) = make_population(5, &dataset, &options, POP_SIZE, 20);
+    let rng = Rng::with_seed(99);
+
+    BestOfSampleBenchEnv {
+        rng,
+        pop: population,
+        stats,
+        options,
+    }
+}
+
+pub fn run_best_of_sample(env: &mut BestOfSampleBenchEnv) {
+    let _ = crate::selection::best_of_sample(&mut env.rng, &env.pop, &env.stats, &env.options);
+}
+
+pub struct NextGenerationBenchEnv {
+    dataset: Dataset<BenchT>,
+    pop: crate::population::Population<BenchT, BenchOps, BENCH_D>,
```

Comment:
```suggestion
    pop: Population<BenchT, BenchOps, BENCH_D>,
```
clean up (see CONTRIBUTORS.md - multiple uses of a common type means we can import)